File: Roadmap/04 - CADENAS DE CARACTERES/python/juanppdev.py
# Definir una cadena de caracteres
cadena = "Hola, mundo!"

# Acceso a caracteres específicos
primer_caracter = cadena[0]
ultimo_caracter = cadena[-1]

print("Acceso a caracteres:")
print("Primer caracter:", primer_caracter)
print("Último caracter:", ultimo_caracter)

# Subcadenas
subcadena = cadena[0:4]
print("\nSubcadena:")
print("Subcadena de los primeros 4 caracteres:", subcadena)

# Longitud de la cadena
longitud = len(cadena)
print("\nLongitud de la cadena:")
print("Longitud de la cadena:", longitud)

# Concatenación
otra_cadena = " ¡Python es genial!"
cadena_concatenada = cadena + otra_cadena
print("\nConcatenación:")
print("Cadena concatenada:", cadena_concatenada)

# Repetición
cadena_repetida = cadena * 3
print("\nRepetición:")
print("Cadena repetida 3 veces:", cadena_repetida)

# Recorrido
print("\nRecorrido:")
for caracter in cadena:
    print(caracter, end=" ")

# Conversión a mayúsculas y minúsculas
print("\n\nConversión a mayúsculas y minúsculas:")
mayusculas = cadena.upper()
minusculas = cadena.lower()
print("Mayúsculas:", mayusculas)
print("Minúsculas:", minusculas)

# Reemplazo
cadena_reemplazada = cadena.replace("mundo", "Python")
print("\nReemplazo:")
print("Cadena con reemplazo:", cadena_reemplazada)

# División
palabras = cadena.split(",")
print("\nDivisión:")
print("Cadena dividida por coma:", palabras)

# Unión
nueva_cadena = "-".join(palabras)
print("\nUnión:")
print("Palabras unidas por guión:", nueva_cadena)

# Interpolación
nombre = "Juan"
edad = 25
mensaje = f"Hola, me llamo {nombre} y tengo {edad} años."
print("\nInterpolación:")
print("Mensaje interpolado:", mensaje)

# Verificación
es_digito = cadena.isdigit()
print("\nVerificación:")
print("¿La cadena contiene solo dígitos?:", es_digito)

## Ejercicio Extra
def es_palindromo(palabra):
    reversa = palabra[::-1]
    return palabra == reversa

def es_anagrama(palabra1, palabra2):
    return sorted(palabra1) == sorted(palabra2)

def es_isograma(palabra):
    set_de_caracteres = set()
    for caracter in palabra:
        if caracter in set_de_caracteres:
            return False  # Si se encuentra un carácter repetido, no es un isograma
        set_de_caracteres.add(caracter)
    return True

# Ingresa las dos palabras que deseas analizar
palabra1 = input("Ingresa la primera palabra: ").lower()
palabra2 = input("Ingresa la segunda palabra: ").lower()

# Comprobación de palíndromo
es_palindromo_1 = es_palindromo(palabra1)
es_palindromo_2 = es_palindromo(palabra2)

# Comprobación de anagrama
es_anagrama_resultado = es_anagrama(palabra1, palabra2)

# Comprobación de isograma
es_isograma_1 = es_isograma(palabra1)
es_isograma_2 = es_isograma(palabra2)

# Mostrar resultados
print("\nResultados:")
print(f"Palabra 1 es palíndromo: {es_palindromo_1}")
print(f"Palabra 2 es palíndromo: {es_palindromo_2}")
print(f"Son anagramas: {es_anagrama_resultado}")
print(f"Palabra 1 es isograma: {es_isograma_1}")
print(f"Palabra 2 es isograma: {es_isograma_2}")

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/Roadmap/04 - CADENAS DE CARACTERES/kotlin/juanppdev.kt
fun main() {
    // Definir una cadena de caracteres
    val cadena = "Hola, mundo!"

    // Acceso a caracteres específicos
    val primerCaracter = cadena[0]
    val ultimoCaracter = cadena[cadena.length - 1]

    println("Acceso a caracteres:")
    println("Primer caracter: $primerCaracter")
    println("Último caracter: $ultimoCaracter")

    // Subcadenas
    val subcadena = cadena.substring(0, 4)
    println("\nSubcadena:")
    println("Subcadena de los primeros 4 caracteres: $subcadena")

    // Longitud de la cadena
    val longitud = cadena.length
    println("\nLongitud de la cadena:")
    println("Longitud de la cadena: $longitud")

    // Concatenación
    val otraCadena = " ¡Kotlin es genial!"
    val cadenaConcatenada = cadena + otraCadena
    println("\nConcatenación:")
    println("Cadena concatenada: $cadenaConcatenada")

    // Repetición
    val cadenaRepetida = cadena.repeat(3)
    println("\nRepetición:")
    println("Cadena repetida 3 veces: $cadenaRepetida")

    // Recorrido
    println("\nRecorrido:")
    for (caracter in cadena) {
        print("$caracter ")
    }

    // Conversión a mayúsculas y minúsculas
    println("\n\nConversión a mayúsculas y minúsculas:")
    val mayusculas = cadena.toUpperCase()
    val minusculas = cadena.toLowerCase()
    println("Mayúsculas: $mayusculas")
    println("Minúsculas: $minusculas")

    // Reemplazo
    val cadenaReemplazada = cadena.replace("mundo", "Kotlin")
    println("\nReemplazo:")
    println("Cadena con reemplazo: $cadenaReemplazada")

    // División
    val palabras = cadena.split(",")
    println("\nDivisión:")
    println("Cadena dividida por coma: $palabras")

    // Unión
    val nuevaCadena = palabras.joinToString("-")
    println("\nUnión:")
    println("Palabras unidas por guión: $nuevaCadena")

    // Interpolación
    val nombre = "Juan"
    val edad = 25
    val mensaje = "Hola, me llamo $nombre y tengo $edad años."
    println("\nInterpolación:")
    println("Mensaje interpolado: $mensaje")

    // Verificación
    val esDigito = cadena.all { it.isDigit() }
    println("\nVerificación:")
    println("¿La cadena contiene solo dígitos?: $esDigito")
}

// Ejercicio Extra
fun main() {
    // Ingresa las dos palabras que deseas analizar
    print("Ingresa la primera palabra: ")
    val palabra1 = readLine()!!.toLowerCase()

    print("Ingresa la segunda palabra: ")
    val palabra2 = readLine()!!.toLowerCase()

    // Comprobación de palíndromo
    val esPalindromo1 = esPalindromo(palabra1)
    val esPalindromo2 = esPalindromo(palabra2)

    // Comprobación de anagrama
    val esAnagrama = esAnagrama(palabra1, palabra2)

    // Comprobación de isograma
    val esIsograma1 = esIsograma(palabra1)
    val esIsograma2 = esIsograma(palabra2)

    // Mostrar resultados
    println("\nResultados:")
    println("Palabra 1 es palíndromo: $esPalindromo1")
    println("Palabra 2 es palíndromo: $esPalindromo2")
    println("Son anagramas: $esAnagrama")
    println("Palabra 1 es isograma: $esIsograma1")
    println("Palabra 2 es isograma: $esIsograma2")
}

fun esPalindromo(palabra: String): Boolean {
    val reversa = palabra.reversed()
    return palabra == reversa
}

fun esAnagrama(palabra1: String, palabra2: String): Boolean {
    return palabra1.toCharArray().sorted() == palabra2.toCharArray().sorted()
}

fun esIsograma(palabra: String): Boolean {
    val setDeCaracteres = HashSet<Char>()
    for (caracter in palabra) {
        if (!setDeCaracteres.add(caracter)) {
            return false // Si se encuentra un carácter repetido, no es un isograma
        }
    }
    return true
}
